File: src/core/health.ts
import { resolve } from 'node:path';
import { getComposeStatus } from '../services/docker.ts';
import {
  COMPOSE_FILE_NAME,
  COMPOSE_PROJECT_NAME,
  HEALTH_CHECK_INTERVAL_MS,
  HEALTH_CHECK_TIMEOUT_MS,
} from '../types/constants.ts';

/** 健康检查结果 */
export interface HealthCheckResult {
  allHealthy: boolean;
  services: ServiceHealth[];
  elapsed: number;
  timedOut: boolean;
}

export interface ServiceHealth {
  name: string;
  state: string;
  healthy: boolean;
}

/**
 * 轮询等待所有服务就绪
 * @param deployDir 部署目录
 * @param onTick 每次轮询的回调，返回当前状态
 */
export async function waitForHealthy(
  deployDir: string,
  onTick?: (services: ServiceHealth[], elapsed: number) => void,
): Promise<HealthCheckResult> {
  const composeFile = resolve(deployDir, COMPOSE_FILE_NAME);
  const startTime = Date.now();

  while (true) {
    const elapsed = Date.now() - startTime;

    // 超时检查
    if (elapsed >= HEALTH_CHECK_TIMEOUT_MS) {
      const statuses = await getComposeStatus(composeFile, COMPOSE_PROJECT_NAME);
      return {
        allHealthy: false,
        services: statuses.map((s) => ({
          name: s.name,
          state: s.state,
          healthy: s.state === 'running' && (s.health === 'healthy' || s.health === ''),
        })),
        elapsed,
        timedOut: true,
      };
    }

    const statuses = await getComposeStatus(composeFile, COMPOSE_PROJECT_NAME);

    const services: ServiceHealth[] = statuses.map((s) => ({
      name: s.name,
      state: s.state,
      // 容器 running 且健康（或无健康检查定义时 health 为空）视为就绪
      healthy: s.state === 'running' && (s.health === 'healthy' || s.health === ''),
    }));

    onTick?.(services, elapsed);

    // 所有服务就绪（且至少有一个服务）
    if (services.length > 0 && services.every((s) => s.healthy)) {
      return {
        allHealthy: true,
        services,
        elapsed,
        timedOut: false,
      };
    }

    // 检查是否有容器退出
    const exited = services.filter((s) => s.state === 'exited' || s.state === 'dead');
    if (exited.length > 0) {
      return {
        allHealthy: false,
        services,
        elapsed,
        timedOut: false,
      };
    }

    // 等待下一次轮询
    await Bun.sleep(HEALTH_CHECK_INTERVAL_MS);
  }
}

File: go-deploy/health.go
package main

import (
	"path/filepath"
	"time"
)

// waitForHealthy polls container health until all services are ready or timeout.
func waitForHealthy(
	deployDir string,
	onTick func(services []ServiceHealth, elapsed int64),
) HealthCheckResult {
	composeFile := filepath.Join(deployDir, ComposeFileName)
	startTime := time.Now()

	for {
		elapsed := time.Since(startTime).Milliseconds()

		// Timeout check
		if elapsed >= HealthCheckTimeoutMs {
			statuses := getComposeStatus(composeFile, ComposeProjectName)
			return HealthCheckResult{
				AllHealthy: false,
				Services:   mapToServiceHealth(statuses),
				Elapsed:    elapsed,
				TimedOut:   true,
			}
		}

		statuses := getComposeStatus(composeFile, ComposeProjectName)
		services := mapToServiceHealth(statuses)

		if onTick != nil {
			onTick(services, elapsed)
		}

		// All services ready (and at least one service exists)
		if len(services) > 0 && allHealthy(services) {
			return HealthCheckResult{
				AllHealthy: true,
				Services:   services,
				Elapsed:    elapsed,
				TimedOut:   false,
			}
		}

		// Any container has exited or died — fail fast
		for _, s := range services {
			if s.State == "exited" || s.State == "dead" {
				return HealthCheckResult{
					AllHealthy: false,
					Services:   services,
					Elapsed:    elapsed,
					TimedOut:   false,
				}
			}
		}

		time.Sleep(HealthCheckIntervalMs * time.Millisecond)
	}
}

// mapToServiceHealth converts ContainerStatus slice to ServiceHealth slice.
// A container is considered healthy when it is running AND either has a
// passing health-check ("healthy") or has no health-check defined ("").
func mapToServiceHealth(statuses []ContainerStatus) []ServiceHealth {
	services := make([]ServiceHealth, 0, len(statuses))
	for _, s := range statuses {
		healthy := s.State == "running" && (s.Health == "healthy" || s.Health == "")
		services = append(services, ServiceHealth{
			Name:    s.Name,
			State:   s.State,
			Healthy: healthy,
		})
	}
	return services
}

// allHealthy returns true when every service in the slice is healthy.
func allHealthy(services []ServiceHealth) bool {
	for _, s := range services {
		if !s.Healthy {
			return false
		}
	}
	return true
}
